### ISSUE TITLE
Should the best practice be to prefix Component and Directive selectors, but not properties of components.

### ISSUE THREAD

**Issue by mhevery** [angular/angular#1598]
## Problem

Assume the following HTML snippet
```html
<button title="OK" tooltip="When done." (click)="..." *if="...">text</button>
```

1. It looks like we are using browser `button`
2. It looks like `title` and `tooltip` are properties of `button`.

### Surprise

Now imagine that the actual implementation is:

```javascript
@Component({ selector:'button', properties: ['title'] })
@View(...)
class Button {
  title: string;
}

@Directives({selector:'[tooltip]', properties: ['tooltip']})
class Tooltip {
  tooltip: string;
}
```

But after seeing the implementation the developer realizes that:

1. `button` is not a native component
2. `tooltip` is a separate, unrelated directive which decorates the `button`

This creates a level of surprise for the developer and effects the readability of the HTML.

## Proposal

As a best practice (not enforced) recommend that:

1. All components are prefixed (but not their properties)
2. All directives are prefixed

## Implication

Now let's rewrite the example with the above rules.

```html
<md-button title="OK" jq-tooltip="When done." (click)="..." *ng-if="...">text</button>
```

It is immediately clear that:

1. `md-button` is a custom component and `title` is its property
2. `jq-tooltip` is a separate directive. Given that they have different prefixes (`md-` and `jq-`) it is unlikely that they have a communication channel. 
3. `ng-if` comes from yet separate library. (again unlikely that they somehow communicate) 

But why prefix the `ng-if`? Well remember that it is short hand for:

```html
<template ng-if="...">
  <md-button title="OK" jq-tooltip="When done." (click)="...">text</button>
</template>
```

And if we don't prefix it it would look like that `template` has `if` property, which is not the case.

**Comment by jelbourn** [comment]
Prefixing for tagName is a *must*. It would be very painful to overwrite the behavior of any native element such as `<button>`.

Something that's important to remember is that `@Component` selector's need not be an element (tag name) selector. As an example, the `MdButton` I currently have built is written:
```html
<button md-button (click)="save()">Save</button>
```
This is done so that all of the standard button behaviors work without any extra work. Does the proposed scheme still seem right in your mind when you have this?
```html
<button md-button no-ripple save="save()">Save</button>
```

Personally I very much like prefixing the Angular-provided directives w/ `ng`, but it's hard to tell how much of that comes from familiarity.

**Comment by mzgol** [comment]
Prefixing tag names would resonate with the Web Component rule that custom components must contain a dash; the rule that was created exactly to prevent conflicts with native elements so I agree it's important.

@jelbourn I think if a `@Component`'s selector is matching an attribute then its properties should be prefixed as well, otherwise we risk conflicts with future new native attributes of special tags like `<button>`.

**Comment by mhevery** [comment]
I think we should have a best practice that components selectors should be on element names. Otherwise @jelbourn  example looks like a directive and as such I would not expect it to have a `@View` (which I think it does not since it is reusing native component).

**Comment by jelbourn** [comment]
@mhevery Without a different API to say "actually render as this other element", using native button element is necessary to make the `md-button` behave just like a regular button.

**Comment by mhevery** [comment]
Could the best practice be to have:
```HTML
<button is="md-button" (click)="save()">Save</button>
```
This way no one is confused that it is not a component.

**Comment by gdi2290** [comment]
I like the idea of enforcing the namespace since it's least surprised for the developer but consider `for-of` microsyntax

```html
<li *ng-for="var item ng-of items | async">
  {{ item }} 
</li>
```

**Comment by gkalpak** [comment]
I also wonder what should be the convention for decorator directives with extra properties, e.g.:

```html
<md-button title="..." jq-tooltip="..."></md-button>
```

Is `title` a property of md-button or jq-tooltip (or none) ?

Is there any good practice for making it clear (and not getting too verbose) ?

**Comment by PascalPrecht** [comment]
I say yes to namespacing component names as a best practice for the same reasons that @mzgol mentioned.

I don't think we'd need to prefix viewport directives, since those are also prefixed with `*`, so it should be quite clear.

Decorator directives should IMO also be namespaced. Because of the same rules that apply to element names: we don't want to conflict with future element attributes that may be introduced.

Regarding @jelbourn's extending existing elements use-case, I'd also vote for `is="component-name"`, since it also streamlines with the custom elements spec to extend existing elements.

The only problem I see here is that, AFAIK `is="component"` in the world of native web components only works on native elements. Whereas Angular directives have always worked on any element/directive/component. Which means, all of a sudden `is="component"` could also work on non-native elements when being in an Angular 2 app.

Regarding property/attribute names.... I really don't know. Having a name space on everything forces us to type a lot. In some cases we'd end up in constructs like:

```html
<foo-component
  foo-component-omg=""
  foo-component-other-omg=""
  foo-component-i-cant-even=""
  ></foo-component>
```

On the other hand, not prefixing everything would be inconsistent...

**Comment by jvandemo** [comment]
#### Proposal
Recommend to prefix anything other than `[properties]` and `(events)`:
- prefix components: `<md-button>`
- prefix properties: `<md-button md-color="red">`
- prefix directives: `<button md-tooltip="...">`

#### Main argument

The new templating syntax of Angular 2 aims at providing the developer with a clear understanding of what is happening by only looking at the markup and without having to dig in the code.

Introducing ambiguities for shorter syntax would work against that goal so best practices should ideally recommend to prefix everything.

#### Pro's
- prevent unexpected behavior as in @mhevery's examples
- prevent collisions between components that use the same properties
- template designers (who are not familiar with Angular but need to style the pages) can better judge what they are styling
- better in line with Web Components standard for components (standard enforces dash in component names)
- maybe even tools like IDE's can take advantage of this by making pieces of the markup clickable and open the source of the component's script
- more future proof in case new attributes or elements are introduced in HTML

#### Cons
- output becomes more verbose so download size of the template would be larger

#### Best practice only

This should be a best practice only and not be enforced in case the developer really wants to target existing HTML elements or attributes.

#### Angular in "strict" mode

Wat if Angular could offer some kind of strict mode where everything without a prefix is not processed?

Suppose your application relies on a third party library and a new version of the library comes out. The new version of the library targets a non-prefixed element. You update the library and suddenly weird things start happening in some parts of your application. This could be really hard to debug, especially in edge cases that may not be visible instantly.

Ideally the developer should check this beforehand but realistically in a large application it is very hard to check the code of all third party libraries during each update.

So to avoid surprises, enabling `strict` mode could assure you that this would never be possible.

**Comment by timjacobi** [comment]
It think @jvandemo 's proposal is a good baseline. Here are some comments and additions.

#### Strict mode
I don't like the idea of a *strict mode*. Why not just get it right the first time and have everyone following best practices from the beginning. The amount of possibilities you have with Angular 1.x are a blessing and a curse and I feel like there is almost room for a *Angular 1.x - The Good Parts* book.

#### Prefixes
I think prefixing is the right way to go.

#### Directive Properties
I don't think prefixing the directive selector *and* the properties is sufficient. Let's have look at an example.
```HTML
<ANY my-directive my-property="value" />
```
It is clear that *my-directive* is the directive but *my-property* is a bit ambiguous. It restricts the directive author in a way that he can't put a directive with the name *my-property* in the *my* namespace without running into issues. I propose a syntax where properties of a directive are prefixed with the fully qualified directive name.
```HTML
<ANY my-directive my-directive-property="value" />
```
This allows us to get rid of the *properties* array in the *@Directive* annotation since it is now inferred by the name.

#### A crazy idea
Now I'm entering experimental, not well thought through territory. How about allowing the user to *rename* the directive in the *@View* annotation?
##### Proposed syntax
```JavaScript
import {If} from 'angular2/angular2';
import {DirectiveWithUnwantedName} from 'wherever';

@View({
    template: `<div yay-better yay-better-property="value" />`,
    directives: {
        'yay-better': DirectiveWithUnwantedName,
        If // using es6 short notation to not rename the "If" directive 
    }
})
```
This could even lead to tools that rename directives to save bytes in the html code (havent thought about the CSS implications yet).

e.g. *some-annoyinglongdirectivename* could be renamed to *x-aldn*

**Comment by pkozlowski-opensource** [comment]
+1 for _requiring_ prefix for directives (both element and decorator) - this would allow us to be closer to web components and do better job at error detection (?)

Not sure I've got opinion on prefixing properties...

**Comment by pkozlowski-opensource** [comment]
@timjacobi in your "crazy idea" by "renaming" you mean "overriding a selector", right? If so I agree that it is needed but I'm not sure it is directly related to the discussion at hand. I would suggest opening a separate issue for this (we might even have one already....)

**Comment by pkozlowski-opensource** [comment]
Oh, actually I do have some opinion on prefixing properties :-) I don't think we should have `<ANY my-directive my-property="value">` - while explicit and unambiguous it would require lots of typing and would make HTML harder to read (?), thus voiding all the benefits.

**Comment by Pajn** [comment]
Renaming directive properties fixes cases like `<ANY my-directive my-property="value" />` however I don't think component properties should be prefixed. It's already clear that unprefixed properties belongs to the component and
 ```
<foo-component
  foo-component-omg=""
  foo-component-other-omg=""
  foo-component-i-cant-even=""
  ></foo-component>
```
It's not just verbose, I do also find it extremely hard to read as my eyes gets lost in all that repeatedness.

So what I would prefer is

-  prefix components: `<md-button>`
-  prefix properties: `<md-button color="red">`
-  prefix directives: `<button md-tooltip="...">`

That would also make it clear what is another directive and what is a property in `<md-button color="red" md-tooltip="...">`

**Comment by jvandemo** [comment]
I think the main issue with not prefixing properties is that different directives applied to the same element could use the same property.

##### Non-prefixed properties are ambiguous

`<any directive-one directive-two color="red">`

what does the `color` property apply to? We cannot know by looking at the markup.

##### Prefixing properties solves the problem but becomes too long to read

`<any directive-one directive-two directive-one-color="red" directive-two-color="blue">`

is perfectly understandable, but becomes too long to read.

##### Alternative shorter syntax that solves both

An alternative more compact syntax could be to pass properties as an object to the directive in case the developer does not want to use prefixes:

`<any directive-one="{ color: 'red' }" directive-two="{ color: 'blue' }">`

which is a lot shorter, reads easily and offers equal semantics.

##### In summary

The proposal for best practice for properties would then be to

prefix properties:
`<any directive-one directive-two directive-one-color="red" directive-two-color="blue">`

OR

pass properties as an object if you don't want to prefix them:
`<any directive-one="{ color: 'red' }" directive-two="{ color: 'blue' }">`

Would that make sense?

**Comment by timjacobi** [comment]
Yet another syntax.
#### Assign properties to directives based on their position on the tag
```HTML
<ANY
    directive-one
    color="red"
    size="5"
    directive-two
    color="blue"
    config="model.config"
/>
```
I think that makes it quite readable. A further addition could be to *require* a dash in the directive name and *disallow* it for properties to avoid name clashes between properties and directives.

##### Alternative
Prefix properties with a symbol
<ANY
    directive-one
    -color="red"
    -size="5"
    directive-two
    -color="blue"
    -config="model.config"
/>
```

**Comment by pkozlowski-opensource** [comment]
@timjacobi do we have any guarantees from the DOM API on the iteration order over attrs?

**Comment by timjacobi** [comment]
@pkozlowski-opensource I'm not familiar enough with the DOM API to answer that but it's worth investigating.

**Comment by jvandemo** [comment]
@timjacobi &mdash; I think there may be 2 possible issues with your first approach:

1. the [HTML 5 spec](http://www.w3.org/html/wg/drafts/html/master/dom.html#global-attributes) ignores duplicates:
`<any class="one" class="two">` will result in `<any class="one">` where the second one is ignored. For some cases this may be an issue.

2. some templating languages offer constructs that do not allow you to determine the exact order of the attributes. For example jade has [`&attributes`](http://jade-lang.com/reference/attributes/) that converts an object to a set of attributes in the resulting HTML but you can't control the order.

Not sure about your second approach. Is that valid HTML?

**Comment by timjacobi** [comment]
I'm not familiar with how Angular 2 renders HTML so if it relies on HTML not swallowing additional properties with the same name then both syntaxes won't work. If I understand the HTML spec correctly most special characters are allowed. Isn't that how Angular 2 does the [property] syntax anyway?

**Comment by alexpods** [comment]
Here is my two cents: let's introduce some "propsAttribute" which will contain attribute name for directive properties. For example:

```html
<element
    directive-one="
        color: 'red'
        size:  5
    "
    directive-two="
        color: 'blue'
        size:  model.config
    "
    directive-single-line="color: colorValue, size: 15"
/>
```


```js
@Decorator({
    selector: '[directive-one]',
    propsAttribute: 'directive-one' // take directive properties
                                    // from this attribute
})
class DirectiveOne {
    color: string,
    size:  int

    // ...
}

@Decorator({
    selector: 'element', // <- directive may not have attribute name in selector
    propsAttribute: 'directive-two'
})
class DirectiveTwo {
    // ...
}

@Decorator({
    selector: 'element[directive-single-line]'
    // By default class name converted into dash-case is used for "propsAttribute"
})
class DirectiveSingleLine {
    //...
}
```

**Props**
* prevent collisions between components 
* much sorter and cleaner syntax than prefixes

**Cons**
* maybe it can not be realized because it completely contradicts with already implemented directive interface :(

**Comment by ewinslow** [comment]
+1 for @jvandemo latest proposal of passing an object hash. Except I'd argue for just standardizing on using the pass-an-object approach and eliminate the prefixed-extra-properties approach. Why bother allowing both? Pick one so it's consistent across teams.

Plus, if you have `<span [some-dir]="stuff" [some-dir-foo]="otherStuf"></span>` you can't tell very well whethere `some-dir-foo` configures the property "foo" of the `some-dir` directive, or whether it is it's own directive entirely. Using an object hash for configuration makes this unambiguous always. Just don't let  directives add extra properties...

**Comment by jpsfs** [comment]
Hi!

@timjacobi, I think the rename of a component (by overriding a selector) is a must for enterprise grade applications.

In my opinion, the composition pattern is a good approach when dealing with small applications but hell breaks loose when you start to think about massive enterprise grade apps. Keeping components loosely coupled is not looking easy on Angular2, and this simple feature can be a great improvement on that.

@pkozlowski-opensource do you know if there is any issue open for this? I wasn't able to find one.

Cheers!

**Comment by PascalPrecht** [comment]
@mhevery could you add some transparency on why `ng-` is re-introduced again?

**Comment by jvandemo** [comment]
@mhevery &mdash; great call on prefixing the directives, makes a lot of sense.

However, as @PascalPrecht noticed, is there a consensus on handling properties yet?

To prefix or not to prefix? Or pass as object?

Would be great to hear your thoughts instead of letting this thread (and valuable inputs from multiple people) die without some kind of best practice. Thanks!!

**Comment by timjacobi** [comment]
+1
Would be nice to know what the conclusion of this is.

**Comment by cssensei** [comment]
I think prefixing it's the best approach just to avoid confusion and leave some space for a lighter syntax instead of passing a config object.

**Comment by timkindberg** [comment]
My idea was exactly the same as @jvandemo of passing an object hash and don't allow decorator directives to access other attributes.

**Comment by iambastardo** [comment]
clashes is inevitable, maybe below???

<namespace.component directive1='' directive2=''/>

**Comment by amypellegrini** [comment]
Some advice/disclaimer in the docs would be nice, I started to create my own components today and my first thought was "what happens if my component name collides with a tag name?" and I ended up here.

**Comment by zoechi** [comment]
@amypellegrini You control what you add to `imports` and `declarations` in `@NgModule()`. 
which should limit the likelyhood of collisions. Using a custom prefix might still be a good idea though.

**Comment by fiznool** [comment]
Re: https://github.com/angular/angular/issues/1598#issuecomment-99890076

> I think the rename of a component (by overriding a selector) is a must for enterprise grade applications.

Was anything ever implemented so that components could be registered with a configurable selector? I searched the docs but couldn't find anything.
